identity/app/controllers/AccountDeletionController.scala

```diff
@@ -0,0 +1,162 @@
+      case object FailedToUnauthenticateUser extends AccountDeletionFailures
+      case object FailedToDeleteAccount extends AccountDeletionFailures
+      case object FailedToRemoveFromAllEmailLists extends AccountDeletionFailures
+
+      def checkUserEnteredCorrectPassword(): EitherT[Future, AccountDeletionFailures, Unit] =
+        (for {
+          auth <- EitherT.fromEither(identityApiClient.authBrowser(emailPasswdAuth, idRequest.trackingData))
+          _ <- EitherT.fromEither(signinService.getCookies(Future.successful(Right(auth)), true))
+        } yield ()).leftMap(_ => FailedToEnterCorrectPassword)
+
+      def unauthenticateUser(): EitherT[Future, AccountDeletionFailures, CookiesResponse] =
+        EitherT.fromEither(api.unauth(emailPasswdAuth, idRequest.trackingData)).leftMap(_ => FailedToUnauthenticateUser)
+
+      def removeFromAllMalingLists(): EitherT[Future, AccountDeletionFailures, ETSubscriber] =
+        EitherT(exactTargetService.unsubscribeFromAllLists(emailPasswdAuth.email)).leftMap(_ => FailedToRemoveFromAllEmailLists)
+
+      def deleteAccountProper(): EitherT[Future, AccountDeletionFailures, Unit] =
+        EitherT.fromEither(identityApiClient.deleteAccount(request.user.auth, emailPasswdAuth)).leftMap(_ => FailedToDeleteAccount)
+
+      def clearCookiesAndDisplaySuccessForm() =
+        NoCache(SeeOther(routes.AccountDeletionController.renderAccountDeletionConfirmForm().url))
+          .discardingCookies(cookiesToDiscard: _*)
+
+      def processFailures(failure: AccountDeletionFailures) = failure match {
+          case FailedToEnterCorrectPassword =>
+            SeeOther(routes.AccountDeletionController.renderAccountDeletionForm ().url)
+              .flashing(boundForm.withError ("password", "Password is incorrect").toFlash)
+
+          case FailedToUnauthenticateUser =>
+            logger.error(s"Failed to un-authenticate user ${request.user.user.id} during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToRemoveFromAllEmailLists =>
+            logger.error(s"Failed to remove user ${request.user.user.id} from all mailing lists during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToDeleteAccount =>
+            logger.error(s"Failed to delete account for user ${request.user.user.id}")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+        }
+
+      (for {
+        _ <- checkUserEnteredCorrectPassword()
+        _ <- removeFromAllMalingLists()
+        _ <- unauthenticateUser()
+        _ <- deleteAccountProper()
+      } yield ()).fold(processFailures, _ => clearCookiesAndDisplaySuccessForm)
+    }
+
+    sealed trait DeletionType
+    case object AutoDeletion extends DeletionType
+    case class ManualDeletion(hasSavedArticles: Boolean, hasComment: Boolean, hasJob: Boolean, isSubscriber: Boolean, isMember: Boolean) extends DeletionType
```

Comment:
Would this be simpler as a `case class DeletionType` with all the fields and a method (or function elsewhere) that tests if it should be auto/manual? Means the logic for testing that would be in one place which is relevant if you expect the list of conditions to grow (and it sounds like that is the case).

There isn't an obvious test for this logic which is a shame given how important it is. By putting the logic in one place (and outside the controller) it'd be easy to add a unit test.